### PR TITLE
Seamless slices str

### PR DESCRIPTION
**Pull request by bhansconnect**

### CODE REVIEW

File: crates/compiler/builtins/bitcode/src/str.zig [hunk 1]
```diff
@@ -57,18 +58,34 @@ pub const RocStr = extern struct {
         return result;
     }
 
-    pub fn fromByteList(list: RocList) RocStr {
-        // TODO: upon adding string seamless slices, I believe this branch can be changed to bit manipulation.
+    // This requires that the list is non-null.
+    pub fn fromSubList(list: RocList, start: usize, count: usize, update_mode: UpdateMode) RocStr {
+        const start_byte = @ptrCast([*]u8, list.bytes) + start;
```
Comment (folkertdev) [inline crates/compiler/builtins/bitcode/src/str.zig:63]: we don't seem to be doing bounds checks here. is that OK?

File: crates/compiler/builtins/bitcode/src/str.zig [hunk 2]
```diff
@@ -129,17 +176,18 @@ pub const RocStr = extern struct {
...
-            }
-        }
+        // const self_words = @ptrCast([*]const usize, @alignCast(@alignOf(usize), self_bytes));
+        // const other_words = @ptrCast([*]const usize, @alignCast(@alignOf(usize), other_bytes));
+        // var w: usize = 0;
+        // while (w < self_len / @sizeOf(usize)) : (w += 1) {
+        //     if (self_words[w] != other_words[w]) {
+        //         return false;
+        //     }
+        // }
```
Comment (folkertdev) [inline crates/compiler/builtins/bitcode/src/str.zig:186]: clean this up?

File: crates/compiler/builtins/bitcode/src/str.zig [hunk 3]
```diff
@@ -129,17 +176,18 @@ pub const RocStr = extern struct {
...
 
         // Compare the leftover bytes
-        var b = w * @sizeOf(usize);
+        // var b = w * @sizeOf(usize);
+        var b: usize = 0;
```
Comment (folkertdev) [inline crates/compiler/builtins/bitcode/src/str.zig:190]: cleanup?

File: crates/compiler/builtins/bitcode/src/str.zig [hunk 4]
```diff
@@ -811,13 +864,40 @@ pub fn strSplit(string: RocStr, delimiter: RocStr) callconv(.C) RocList {
     return list;
 }
 
+const Init = fn (slice_bytes: [*]u8, len: usize, ref_ptr: usize) RocStr;
```
Comment (folkertdev) [inline crates/compiler/builtins/bitcode/src/str.zig:867]: this const is unused

File: crates/compiler/builtins/bitcode/src/str.zig [hunk 5]
```diff
@@ -811,13 +864,40 @@ pub fn strSplit(string: RocStr, delimiter: RocStr) callconv(.C) RocList {
...
+fn initFromBigStr(slice_bytes: [*]u8, len: usize, ref_ptr: usize) RocStr {
+    // Here we can make seamless slices instead of copying to a new small str.
+    return RocStr{
+        .str_bytes = slice_bytes,
+        .str_len = len | SEAMLESS_SLICE_BIT,
+        .str_capacity = ref_ptr,
+    };
+}
+
+// TODO: relpace this with @qualCast or @constCast in future version of zig
```
Comment (folkertdev) [inline crates/compiler/builtins/bitcode/src/str.zig:882]: will 0.10 have this?

File: crates/compiler/gen_llvm/src/llvm/build_str.rs [hunk 6]
```diff
@@ -63,3 +63,19 @@ pub(crate) fn str_equal<'a, 'ctx, 'env>(
         bitcode::STR_EQUAL,
     )
 }
+
+// Gets a pointer to just after the refcount for a list or seamless slice.
+// The value is just after the refcount so that normal lists and seamless slices can share code paths easily.
+pub(crate) fn str_refcount_ptr<'a, 'ctx, 'env>(
```
Comment (folkertdev) [inline crates/compiler/gen_llvm/src/llvm/build_str.rs:69]: not the best name (the pointer is not to the refcount), but I don't have better suggestions

File: crates/compiler/test_gen/src/helpers/from_wasm32_memory.rs [hunk 7]
```diff
@@ -60,7 +60,7 @@ impl FromWasm32Memory for RocStr {
         let str_words: &[u32; 3] = unsafe { std::mem::transmute(&str_bytes) };
 
         let big_elem_ptr = str_words[Builtin::WRAPPER_PTR as usize] as usize;
-        let big_length = str_words[Builtin::WRAPPER_LEN as usize] as usize;
+        let big_length = str_words[Builtin::WRAPPER_LEN as usize] as usize & (i32::MAX as usize);
```
Comment (folkertdev) [inline crates/compiler/test_gen/src/helpers/from_wasm32_memory.rs:63]: this deserves a comment I think

File: crates/compiler/builtins/bitcode/src/str.zig (hunk 1 above)
Comment (bhansconnect) [reply crates/compiler/builtins/bitcode/src/str.zig]: added comment with more detail and made name clearer.

File: crates/compiler/builtins/bitcode/src/str.zig (hunk 2 above)
Comment (bhansconnect) [reply crates/compiler/builtins/bitcode/src/str.zig]: done

File: crates/compiler/builtins/bitcode/src/str.zig (hunk 4 above)
Comment (bhansconnect) [reply crates/compiler/builtins/bitcode/src/str.zig]: Oh, yeah, I guess I never use the type. Just functions of the type. removed.

File: crates/compiler/builtins/bitcode/src/str.zig (hunk 5 above)
Comment (bhansconnect) [reply crates/compiler/builtins/bitcode/src/str.zig:872]: I believe 0.10 will have `@qualCast` then in 0.11 it is split into `@constCast` and some other cast.

File: crates/compiler/gen_llvm/src/llvm/build_str.rs (hunk 6 above)
Comment (bhansconnect) [reply crates/compiler/gen_llvm/src/llvm/build_str.rs:69]: Yeah, was the same issue with lists. The function was originally actually refcount pointer but then I changed it for performance. Maybe just allocation pointer, or original str pointer?

File: crates/compiler/test_gen/src/helpers/from_wasm32_memory.rs (hunk 7 above)
Comment (bhansconnect) [reply crates/compiler/test_gen/src/helpers/from_wasm32_memory.rs]: done

**Review by folkertdev — APPROVED** [review]

**Pull request merged**